### PR TITLE
add wasm and watch the whole project burn

### PR DESCRIPTION
**Pull request by SaintMercury**
Adds wasm-bindgen and also doesn't work :)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -10,12 +10,16 @@ license = "MIT OR Apache-2.0"
 keywords = ["TES3", "Morrowind"]
 categories = ["Game development"]
 
+[lib]
```
Comment (SaintMercury) [inline Cargo.toml:13]: this might be unnecessary. Can't remember adding it, must've done so in the other repo

File: libs/esp/Cargo.toml [hunk 2]
```diff
@@ -13,7 +13,9 @@ esp_macros = { path = "../esp_macros" }
 hashbrown = { version = "^0.12", features = ["rayon"] }
 rayon = "^1.5"
 smart-default = "^0.6"
+wasm-bindgen = { version = "0.2", optional = true }
 
 [features]
-default = []
+default = ["wasm"]
```
Comment (SaintMercury) [inline libs/esp/Cargo.toml:19]: temp

**Comment by SaintMercury** [comment]
https://github.com/rustwasm/wasm-bindgen/pull/2631
https://github.com/rustwasm/wasm-bindgen/issues/2407

Relevant issues and PR for possible solutions to errors

**Comment by Greatness7** [comment]
I recommend working off the `dev` branch if possible. It's got a lot of changes and improvements. I plan to merge it into main eventually.

Besides that, since `serde` support is already implemented there, I don't think you actually need to decorate every struct with `#[wasm_bindgen]` and can instead just let [serde_wasm_bindgen](https://crates.io/crates/serde-wasm-bindgen) do it for you. 

Example:

```toml
[dependencies]
wasm-bindgen = "^0.2"
serde = "^1.0"
serde-wasm-bindgen = "^0.6"
js-sys = "^0.3"
web-sys = "^0.3"

[dependencies.tes3]
git = "https://github.com/Greatness7/tes3"
branch = "dev"
default-features = false
features = ["esp", "nightly", "serde"]
```

```rs
use js_sys::Uint8Array;
use serde_wasm_bindgen::{from_value, to_value};
use tes3::esp::Plugin;
use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn load_objects(array: Uint8Array) -> Result<JsValue, JsValue> {
    let mut plugin = Plugin::new();
    plugin
        .load_bytes(array.to_vec().as_ref())
        .map_err(|e| JsValue::from(e.to_string()))?;

    let value = to_value(&plugin.objects)?;

    Ok(value)
}

#[wasm_bindgen]
pub fn save_objects(value: JsValue) -> Result<Uint8Array, JsValue> {
    let mut plugin = Plugin {
        objects: from_value(value)?,
    };

    let bytes = plugin
        .save_bytes()
        .map_err(|e| JsValue::from(e.to_string()))?;

    let length = u32::try_from(bytes.len()) //
        .map_err(|e| JsValue::from(e.to_string()))?;

    let array = Uint8Array::new_with_length(length);
    array.copy_from(&bytes);

    Ok(array)
}
```

**Pull request closed without merging**